src/asyncresults.jl

```diff
@@ -0,0 +1,242 @@
+
+This implements the loop described in the PostgreSQL documentation for
+[Asynchronous Command Processing](https://www.postgresql.org/docs/10/libpq-async.html).
+
+The `throw_error` option only determines whether to throw errors when handling the new
+[`Result`](@ref)s; the `Task` may error for other reasons related to processing the
+asynchronous loop.
+
+The result returned from the `Task` will be the [`Result`](@ref) of the last query run (the
+only query if using parameters).
+Any errors produced by the queries will be thrown together in a `CompositeException` by
+`@sync`.
+"""
+function consume(async_result::AsyncResult; throw_error=true)
+    @async begin
+        try
+            debug(LOGGER, "getting the socket fd")
+            pqfd = socket(async_result.jl_conn)
+            try
+                debug(LOGGER, "making an FDWatcher")
+                watcher = FDWatcher(pqfd, true, false)  # readable, not writeable
+                try
+                    last_result = @sync begin
+                        result = nothing
+                        debug(LOGGER, "beginning the waiting loop")
+                        while true
+                            debug(LOGGER, "waiting on the fd")
+                            wait(watcher)
+                            debug(LOGGER, "consuming input")
+                            success = libpq_c.PQconsumeInput(async_result.jl_conn.conn) == 1
+                            !success && error(error_message(async_result.jl_conn))
+                            debug(LOGGER, "checking isbusy")
+                            while libpq_c.PQisBusy(async_result.jl_conn.conn) == 0
+                                debug(LOGGER, "getting a result")
+                                result_ptr = libpq_c.PQgetResult(async_result.jl_conn.conn)
+                                result_ptr == C_NULL && @goto finished
+                                debug(LOGGER, "handling the non-null result")
+                                result = @async handle_result(Result(
+                                    result_ptr, async_result.jl_conn;
+                                    async_result.result_kwargs[]...
+                                ); throw_error=throw_error)
+                            end
+                        end
+                        @label finished
+                        debug(LOGGER, "finished the sync block")
+                        result
+                    end
+
+                    return fetch(last_result)
+                finally
+                    close(watcher)
+                end
+            finally
+                # I should maybe close pqfd here but I don't think that's possible since
+                # it's a file descriptor not a handle and it's not technically open
+            end
+        finally
+            # ensure the connection is unlocked
+            _ensure_unlocked(async_result)
+            debug(LOGGER, "finished the async block")
+        end
+    end
+end
```

Comment:
Can we break this up into several functions?
That would also let us get rid of the `@goto` for escaping multiple levels of the loop,
since `return` could be used instead

Reply:
Unfortunately `@sync` cares about lexically-enclosed `@async` now, so they have to be together. This means that won't help us get rid of the `@goto`.

Reply:
Maybe we should just not use `@sync` then. And just use `wait`/`fetch` on the returned `Task` as required

Reply:
Maybe we should just not use `@sync` then. And just use `wait`/`fetch` on the returned `Task` as required

Reply:
Maybe we should just not use `@sync` then. And just use `wait`/`fetch` on the returned `Task` as required

Reply:
Maybe we should just not use `@sync` then. And just use `wait`/`fetch` on the returned `Task` as required

Reply:
Maybe we should just not use `@sync` then. And just use `wait`/`fetch` on the returned `Task` as required

Reply:
`@sync` handles bundling every error that occurs in the enclosed `@async`s into a `CompositeException`.

Reply:
I am not convinces we shouldn't write some code that does this via some function

```
function mysync(f)
    futurish = try
        return f()
    catch err
        return Future(CompositeException(err)
    end 
    return fetch(futurish_thing)
end
```
``

Reply:
You need the errors from every result though. 

`PQsendQuery` can execute many statements. Each of those statements returns a result from `PQgetResult`. You *must* call `PQgetResult` for each statement. Each of those statements may error. Each of those statements may return results. I have chosen to only return the result for the last statement, as that's what `libpq`'s synchronous API does (`PQexec`). However, I want to throw every error that occurs. So, for each `PQgetResult`, I'd need to call `handle_result` and either catch the error and append that error to a list or store the result in a variable.

After thinking about this for a while I think I could just do the loop that collects all the result pointers and then process them after the loop. Then I could move the loop into a function. I'll do that tomorrow morning.

Reply:
I do not harte this any more